### PR TITLE
Server

### PR DESCRIPTION
**Pull request by naota-nandaba**
- folder structure/set up
- models
- routes for posts
- scraper func
- test route to add user

### CODE REVIEW

**Comment by naota-nandaba** [comment]
@OfTheDelmer

File: server/index.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+// const cookieSession = require('cookie-session');
+
+const bodyParser = require('body-parser');
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({extended: false}));
+
+const request = require('request');
+const cheerio = require('cheerio');
+
+// app.use(cookieSession({
```
Comment (OfTheDelmer) [inline server/index.js:22]: Remove commented out code.

File: server/index.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+//   name: 'auth_app_user',
+//   keys: ['socure1', 'socure2']
+// }));
+
+//CORS for Client
+app.options('/*', (req, res, next) => {
```
Comment (OfTheDelmer) [inline server/index.js:28]: Would probably change this to be:

```js
const OPTIONS = 'OPTIONS';

// Pull out CORS headers into a constant.
const CORS_HEADERS = 
  Object.freeze({
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type, authorization',
      'Access-Control-Allow-Credentials': 'true',
      'Access-Control-Allow-Methods':  'GET, POST, OPTIONS, PUT, PATCH',
   });

app.use('/*', (req, res, next) => {
  res.set(CORS_HEADERS);

  if (req.method === OPTIONS) {
    // Handling for preflight requests
     res.sendStatus(200); 
  } else  {
     next();
  }
});
```

You could also break it into a middleware and `options` handler.

```js
const ANY_PATH = '/*';

const CORS_HEADERS = 
  Object.freeze({
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type, authorization',
      'Access-Control-Allow-Credentials': 'true',
      'Access-Control-Allow-Methods':  'GET, POST, OPTIONS, PUT, PATCH',
   });

app.use(ANY_PATH, (req, res, next) => {
     res.set(CORS_HEADERS);
     next();
});

app.options(ANY_PATH, (req, res) => { res.sendStatus(200); })
```


Good reading on the CORS handling: [MDN Access Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS)

File: server/index.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  bcrypt.hash(req.body.password, 12)
+    .then(function (digest) {
+      // use knex to insert the digested password
+      knex('users')
+      .insert({
+          email: req.body.email,
+          hashed_password: digest
+        }, '*')
+        .then(function (users) {
+           console.log(users)
```
Comment (OfTheDelmer) [inline server/index.js]: Careful about leaving these console logs.

File: server/index.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+          const user = users[0];
+
+          res.send(`Welcome ${user.email}`)
```
Comment (OfTheDelmer) [inline server/index.js]: This won't really be utilized by the UI so you can just `res.sendStatus(204)`

File: server/index.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        })
+    })
+
+    .catch(function (err) {
+      console.log(err);
+      res.send("Error");
```
Comment (OfTheDelmer) [inline server/index.js]: `res.sendStatus(401)`.

File: server/index.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  var str = "Macaroni"
+
+  var splitStr = str.split(" ");
+
+  splitStr.forEach(function(el){
+    el.replace(/\.$/, " ");
+  })
+
+  var newStr = splitStr.filter(function (el){
+    return el !== 'A'
```
Comment (OfTheDelmer) [inline server/index.js:80]: Moving this into an array and using `indexOf` or using a RegExp could help clean this up

File: server/models/base/relations.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+const UNREGISTERED_TABLE = (name) => `Unregisted relation: ${name}`;
```
Comment (OfTheDelmer) [inline server/models/base/relations.js]: Remove this file because it's not being used.

File: ui/public/css/app.css [hunk 8]
```diff
@@ -1,44 +1,154 @@
+@media (min-width: 500px) {
```
Comment (OfTheDelmer) [inline ui/public/css/app.css]: Duplicate file. Remove from PR.

**Review by OfTheDelmer — CHANGES_REQUESTED** [review]
You'll want to merge your UI first. This server PR needs work, and contains work from UI that you'll want to remove.

**Pull request merged**